hub/core/chunk_engine/generator.py

```diff
@@ -0,0 +1,95 @@
+import numpy as np
+from typing import Generator, Tuple
+
+from hub.core.chunk_engine.exceptions import ChunkGeneratorError
+
+
+def chunk(
+    content_bytes: bytes, previous_num_bytes: int, chunk_size: int
+) -> Generator[Tuple[bytes, int], None, None]:
+    """
+    Generator function that chunks bytes.
+
+    Chunking is the process of taking the input `content_bytes` & breaking it up into smaller chunks,
+    the sizes of which are <= `chunk_size`.
+
+    Example 1:
+        content_bytes = b"101010101010"
+        chunk_size = 4
+        output_chunks = [b"1010", b"1010", b"1010"]
+
+        This is considered a "perfect fit" because all bytes fit perfectly into 3 chunks.
+
+
+    Example 2:
+        content_bytes = b"10101010111122"
+        chunk_size = 4
+        output_chunks = [b"1010", b"1010", b"1111", b"22"]
+
+        As you can see, the last chunk in `output_chunks` only has 2 bytes (chunk_size=4), which means
+        this is not a perfect fit, but rather it has a partial chunk at the end.
+
+
+    Args:
+        content_bytes(bytes): Bytes object with the data to be chunked.
+        previous_num_bytes(int): The number of bytes in the previous chunk. This is helpful for filling the previous chunk before starting a new one.
+        chunk_size(int): Each individual chunk will be assigned this many bytes maximum.
+
+    Yields:
+        Each yield is a chunk of the `content_bytes`. Each chunk is of length (0, `chunk_size`].
+
+    Raises:
+        ChunkGeneratorError: If leftover bytes are negative or the previous chunk was invalid.
+    """
+
+    if len(content_bytes) <= 0:
+        return
+
+    if previous_num_bytes is None:
+        bytes_left = 0
+    else:
+        bytes_left = chunk_size - previous_num_bytes
+
+    content_num_bytes = len(content_bytes)
+
+    if bytes_left < 0:
+        raise ChunkGeneratorError(
+            "Previous chunk exceeded `chunk_size` (%i > %i)." % (bytes_left, chunk_size)
+        )
+
+    # handle filling the rest of the previous chunk
+    total_bytes_yielded = 0
+    if bytes_left > 0:
+        content_bytes_piece = content_bytes[:bytes_left]
```

Comment:
shouldn't content_num_bytes be updated after this? (Not for line 67 but for line 72 onwards)
Ideal solution:-
After line 63 add-
`content_num_bytes -= len(content_bytes_piece)
`
Change line 67 to-
`if content_num_bytes <= 0`

Reply:
no, because `content_bytes_piece` is just a subset of `content_bytes`. 
`content_bytes` never changes.

Reply:
right but in that case shouldn't line 72 consider the length of the subset and not the full length?

Reply:
holy cow, good catch. i changed that line but in a slightly different way. that would have been a very annoying bug to track down in the future!